pkg/authentication/externalidp/aws_client.go

```diff
@@ -0,0 +1,203 @@
+	ctx := context.TODO()
+	identityToken, err := getIdentityToken(ctx, params)
+	if err != nil {
+		return "", err
+	}
+
+	client, err := apigen.NewClientWithResponses(
+		serverEndpoint,
+		apigen.WithHTTPClient(httpClient),
+	)
+	if err != nil {
+		return "", err
+	}
+
+	tokenTTL := int(params.TokenTTL.Seconds())
+	externalLoginInfo := apigen.ExternalLoginInformation{
+		IdentityRequest: map[string]interface{}{
+			"identity_token": identityToken,
+		},
+		TokenExpirationDuration: &tokenTTL,
+	}
+	externalPrincipalLoginResp, err := client.ExternalPrincipalLoginWithResponse(ctx, apigen.ExternalPrincipalLoginJSONRequestBody(externalLoginInfo))
+	if err != nil {
+		return "", err
+	}
+	if externalPrincipalLoginResp == nil || externalPrincipalLoginResp.JSON200 == nil {
+		return "", ErrRetrievingToken
+	}
+	return externalPrincipalLoginResp.JSON200.Token, nil
+}
+
+func getIdentityToken(ctx context.Context, params *AWSIAMParams) (string, error) {
+	cfg, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return "", err
+	}
+	creds, err := cfg.Credentials.Retrieve(ctx)
+	if err != nil {
+		return "", err
+	}
+	if creds.Expired() {
+		return "", ErrAWSCredentialsExpired
+	}
+	stsClient := sts.NewFromConfig(cfg)
+	stsPresignClient := sts.NewPresignClient(stsClient, func(o *sts.PresignOptions) {
+		o.ClientOptions = append(o.ClientOptions, func(opts *sts.Options) {
+			opts.ClientLogMode = aws.LogSigning
+		})
+	})
+
+	presignGetCallerIdentityResp, err := stsPresignClient.PresignGetCallerIdentity(context.Background(), &sts.GetCallerIdentityInput{},
+		sts.WithPresignClientFromClientOptions(sts.WithAPIOptions(setHTTPHeaders(params.TokenRequestHeaders, params.URLPresignTTL))),
+	)
+	if err != nil {
+		return "", err
+	}
+
+	parsedURL, err := url.Parse(presignGetCallerIdentityResp.URL)
+	if err != nil {
+		return "", err
+	}
+
+	queryParams := parsedURL.Query()
+	credentials := queryParams.Get(AWSAuthCredentialKey)
+	splitedCreds := strings.Split(credentials, "/")
+	calculatedRegion := splitedCreds[2]
+	identityTokenInfo := IdentityTokenInfo{
+		Method:             "POST",
+		Host:               parsedURL.Host,
+		Region:             calculatedRegion,
+		Action:             AWSAuthAction,
+		Date:               queryParams.Get(AWSAuthDateKey),
+		ExpirationDuration: queryParams.Get(AWSAuthExpiresKey),
+		AccessKeyID:        creds.AccessKeyID,
+		Signature:          queryParams.Get(AWSAuthSignatureKey),
+		SignedHeaders:      strings.Split(queryParams.Get(AWSAuthSignedHeadersKey), ";"),
+		Version:            queryParams.Get(AWSAuthVersionKey),
+		Algorithm:          queryParams.Get(AWSAuthAlgorithmKey),
+		SecurityToken:      queryParams.Get(AWSAuthSecurityTokenKey),
+	}
+
+	marshaledIdentityTokenInfo, _ := json.Marshal(identityTokenInfo)
+	encodedIdentityTokenInfo := base64.StdEncoding.EncodeToString(marshaledIdentityTokenInfo)
+	return encodedIdentityTokenInfo, nil
```

Comment:
This section is a good candidate to a separate function, it'll be easier to test it, i.e: 

```go
func NewIdentityTokenInfoFromURL(presignedURL string) (*IdentityTokenInfo, string, error)
```

Note: returning both the encoded string output of the identityTokenInfo and the struct itself - it allows to test it properly, writing a test for encoded string is harder if you want to verify specific fields.